fairlearn/metrics/_mfv2.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+from typing import Dict, List, Callable
+
+import pandas as pd
+
+
+class MetricFunctionRequest:
```

Comment:
Instead of this, we could have something like

``` python
class ConsumerMetric:
    def __init__(self, metric, *, metadata=None):
        self.metric = metric
        self.metadata = set(metadata)

    def __call__(self, y_true, y_pred, **kwargs):
        if set(kwargs.keys()) != self.metadata:
            raise TypeError("Passed metadata don't match requested ones!")

        return self.metric(y_true, y_pred, **kwargs)

    def get_metadata_request(self):
        return self.metadata
```

Then we could use the old `MetricFrame`, and all is as before, except that the user can pass something like `ConsumerMetric(metric_func, metadata=["sample_weight"])`, and the `MetricFrame` checks if a given metric has `get_metadata_request` or not, if not, then it only accepts `y_true` and `y_pred`, otherwise it accepts extra metadata. This would simplify the `sample_params` part of the `MetricFrame` quite a bit.